bindings/go/tests/utils.go

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2021, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package goucxtests
+
+import (
+	. "ucx"
+	"unsafe"
+)
+
+const selfEpTag uint64 = ^uint64(0)
+
+func memoryAllocate(entity *TestEntity, size uint64, memoryType UcsMemoryType) unsafe.Pointer {
+	mmapParams := &UcpMmapParams{}
+	mmapParams.Allocate().SetLength(size).SetMemoryType(memoryType)
+
+	result, err := entity.context.MemMap(mmapParams)
+	if err != nil {
+		entity.t.Fatalf("Failed to allocate memory %v", err)
+	}
+
+	entity.mem = result
+
+	memAttr, _ := entity.mem.Query(UCP_MEM_ATTR_FIELD_ADDRESS)
+
+	return memAttr.Address
+}
+
+func createSelfEp(entity *TestEntity) {
+	if entity.selfEp == nil {
+		ucpAddress, err := entity.worker.GetAddress()
+		if err != nil {
+			entity.t.Fatalf("Failed to get address %v", err)
+		}
+		entity.selfEp, err = entity.worker.NewEndpoint((&UcpEpParams{}).SetUcpAddress(ucpAddress))
+		if err != nil {
+			entity.t.Fatalf("Failed to create endpoint %v", err)
+		}
+	}
+}
+
+func memorySet(entity *TestEntity, data []byte) {
+	memAttr, _ := entity.mem.Query(UCP_MEM_ATTR_FIELD_ADDRESS, UCP_MEM_ATTR_FIELD_LENGTH, UCP_MEM_ATTR_FIELD_MEM_TYPE)
+
+	createSelfEp(entity)
+	recvRequest, _ := entity.worker.RecvTagNonBlocking(memAttr.Address, memAttr.Length, selfEpTag, selfEpTag, &UcpRequestParams{
+		MemType: memAttr.MemType,
+	})
+	defer recvRequest.Close()
+
+	stringBytes := CBytes(data)
+	defer FreeNativeMemory((stringBytes))
+	sendReq, _ := entity.selfEp.SendTagNonBlocking(selfEpTag, stringBytes, uint64(len(data)), &UcpRequestParams{
+		MemType:    UCS_MEMORY_TYPE_HOST,
+		MemTypeSet: true,
+	})
+	defer sendReq.Close()
+
+	for recvRequest.GetStatus() == UCS_INPROGRESS {
+		entity.worker.Progress()
+	}
+}
+
+func memoryGet(entity *TestEntity) []byte {
```

Comment:
can we unite common code between memorySet and memoryGet?

Reply:
Common part is already in `createSelfEp(entity)`. Memory set doing send from Go memory to target memory, while memory get doing get from target to go memory

Reply:
maybe it could be a common function doSendRecv which will get src/dest buffers and length?